Join-Path: Add -SkipValidation for -Resolve

vexx32 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

1. Moves -Resolve to its own parameter set, where it is mandatory.
2. Adds a -SkipValidation parameter to the same set, where it is optional.

The -SkipValidation switch allows Join-Path to use `SessionState.Path.GetUnresolvedProviderPathFromPSPath()` instead of `SessionState.Path.GetResolvedPSPathFromPSPath()`. This allows users to get a meaningful, properly-resolved path to items that may not yet exist.

Previously:
```powershell
PS> Join-Path C:\ .\Test\Test
C:\.\Test\Test
PS> Join-Path C:\ .\Test\Test -Resolve
Join-Path : Cannot find path 'C:\Test\Test' because it does not exist.
At line:1 char:1
+ Join-Path c:\ .\Test\Test -Resolve
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (C:\test\test:String) [Join-Path], ItemNotFoundException
+ FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.JoinPathCommand
```

Now:
```powershell
PS> Join-Path C:\ .\Test\Test -Resolve -SkipValidation
C:\Test\Test
```

## Remaining Work Items

* Add Tests

## PR Context

Related: #2993

Ideally, `Convert-Path`, `Join-Path`, and `Resolve-Path` should all eventually have similar capabilities, but one thing at a time, eh? 😄 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

vexx32 committed:
:sparkles: Add -SkipValidation switch to Join-Path -Resolve

vexx32 committed:
:pencil: Fix doc tag for CodeFactor

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/CombinePathCommand.cs:
```diff
@@ -119,15 +129,27 @@ protected override void ProcessRecord()
                     // Resolve the paths. The default API (GetResolvedPSPathFromPSPath)
                     // does not allow non-existing paths.
-                    Collection<PathInfo> resolvedPaths = null;
+                    Collection<string> resolvedPaths = null;
```
What benefits you expect?
I see extra allocations below.
Also with PathInfo it is easy to enhance the cmdlet in future.

vexx32 replied:
It's not so much that reducing this to string has a benefit, it's that the other provider method doesn't give back PathInfo objects to work with.

If there is a way to create PathInfo objects for nonexistent items, I'm all ears.

If not, the other alternative would be to diverge completely here and handle both cases entirely separately, I suppose?

iSazonov replied:
I'd don't mess these two code paths - old and new.

Also we have ` public string GetUnresolvedProviderPathFromPSPath(string path)`

vexx32 replied:
Fair point, I'll play with the structure. :)

vexx32 replied:
@iSazonov that overload doesn't allow us to pass in `CmdletProviderContext`, which means it would not be able to distinguish the different providers, right? Or am I drawing the wrong conclusions there?

iSazonov replied:
Oops... If we use SkipValidation under Resolve we must use GetResolvedPSPathFromPSPath() API in any case. So we need enhance the API to do not validate paths.

iSazonov replied:
There already exist `allowNonexistingPaths` parameter.

vexx32 replied:
I'm not sure I follow. Why do we need to modify the API? At least on the surface it would appear that `GetUnresolvedProviderPathFromPSPath()` performs the same function as `GetResolvedPSPathFromPSPath()`, at least in the context of its use in this cmdlet. After all, `Join-Path` doesn't really use the `PathInfo` object, simply returning the string value from the `PathInfo.Path` property of each resolved path.

The reason such a path must stay _un_resolved (despite the wording of the parameters vs the API) is that if you give it a path segment containing a wildcard, we can't assume that the wildcard refers to only existing items, by the nature of the request.

iSazonov replied:
resolved - here means "resolve wildcards".

vexx32 replied:
Yes, but if a path doesn't exist a wildcard can't be resolved. Should a different term (different switch entirely?) be used for this kind of operation? It's not really resolving wildcards, but it's validating and resolving relative nodes to a single straightforward path according to the PSProvider's path format

iSazonov replied:
> if a path doesn't exist a wildcard can't be resolved.

Can with SkipValidation.   c:\window*\notexisting.txt -> c:\windows\notexisting.txt

vexx32 replied:
Yes, assuming that part of the path actually exists. But if we're skipping validation to allow resolution of non-existent paths, we can't assume wildcarded segments will necessarily exist?

E.g., if path `C:\test\` exists with no subfolders and user tries `Join-Path -Path C:\Test ..\Test\*\test2 -Resolve -SkipValidation` should the cmdlet still return `C:\Test\*\Test2`, or should it return nothing as it does currently with -Resolve?

iSazonov replied:
I still expect that allowNonexistingPaths is that resolve your questions. Note that current behavior is (1) resolve paths (with wildcards), (2) check existence - we need only turn off second (with allowNonexistingPaths ).

vexx32 replied:
And if the path segment that contains the wildcard isn't found? Just return nothing?

I'll look into adding an overload for the provider method there. 🙂

vexx32 replied:
@iSazonov OK, looks like it was pretty simple overall...

This is the result:
```powershell
PS Z:\Scripts\PowerShell-Core> "c:\aaa" | Join-Path -ChildPath ".\..\bbb" -Resolve
Join-Path : Cannot find path 'C:\bbb' because it does not exist.
At line:1 char:12
+ "c:\aaa" | Join-Path -ChildPath ".\..\bbb" -Resolve
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (C:\bbb:String) [Join-Path], ItemNotFoundException
+ FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.JoinPathCommand
 
PS Z:\Scripts\PowerShell-Core> "c:\aaa" | Join-Path -ChildPath ".\..\bbb" -Resolve -SkipValidation
C:\bbb
PS Z:\Scripts\PowerShell-Core> "c:\aaa" | Join-Path -ChildPath ".\..\bbb\*" -Resolve -SkipValidation
Join-Path : Cannot find path 'C:\bbb' because it does not exist.
At line:1 char:12
+ "c:\aaa" | Join-Path -ChildPath ".\..\bbb\*" -Resolve -SkipValidation
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (C:\bbb:String) [Join-Path], ItemNotFoundException
+ FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.JoinPathCommand
```

In other words, a wildcard is only permitted in a path segment that currently exists, but otherwise nonexistent relative references are resolved accurately and a properly-formed path results.

I _think_ this is OK? Unsure, to be honest,

iSazonov replied:
I think it is right behavior.
Also I'd consider `Join-Path -Path c:\aaa\*\bbb -ChildPath ccc`.

Please contribute and add tests.

vexx32 replied:
🤔 OK, added tests and rebased. That last case you mention where the wildcard may refer to multiple existing paths with non-existing leaf node(s)... that's not handled properly. It just returns nothing at the moment. I'll come back to this later on today or perhaps tomorrow. May need to dig deeper into the underlying methods to see how to best handle that one.

vexx32 committed:
:recycle: :wrench: Refactor method
:wrench: Revert to using PathInfo collection
:recycle: Refactor SkipValidation to its own branch which returns early

vexx32 committed:
:recycle: :sparkles: Add new internal method to handle path resolution
:recycle: Remove unneeded code branch from SplitPathCommand

vexx32 committed:
:recycle: Standardize style in Join-Path tests

vexx32 committed:
:white_check_mark: Added test for -SkipValidation

vexx32 commented:
OK, after digging through this last night, I don't think that what I'm trying to do _can_ be done without a kludge.

I would have to do something along the lines of:

1. Test if the returned paths are null & the input path(s) contain wildcard expressions.
    * Non-wildcard paths should error out rather than returning `null`
2. Attempt to recursively/repeatedly resolve each parent node, splitting off and retaining leaf nodes one at a time, until I obtain a non-null set of paths that resolve to container nodes.
4. Then, I can proceed to iterate over my set of paths, appending the total leaf nodes to them.
5. Return the resulting paths.

Is this an acceptable solution for the behaviour, or is this not worth the change, do you think?
